src/Compilers/CSharp/Portable/Symbols/Tuples/TupleTypeSymbol.cs

```diff
@@ -119,47 +72,32 @@ internal sealed class TupleTypeSymbol : WrappedNamedTypeSymbol
-                          tupleCompatibleType,
-                          elementLocations,
-                          elementNames,
-                          errorPositions);
+            Debug.Assert(tupleCompatibleType.IsTupleType);
+            return tupleCompatibleType.WithElementNames(elementNames, elementLocations, errorPositions, locations);
         }
 
-        public static TupleTypeSymbol Create(ImmutableArray<Location> locations, NamedTypeSymbol tupleCompatibleType,
-            ImmutableArray<Location> elementLocations, ImmutableArray<string> elementNames, ImmutableArray<bool> errorPositions)
+        internal NamedTypeSymbol WithTupleDataFrom(NamedTypeSymbol original)
         {
-            Debug.Assert(tupleCompatibleType.IsTupleCompatible());
-
-            ImmutableArray<TypeWithAnnotations> elementTypes;
-
-            if (tupleCompatibleType.Arity == RestPosition)
+            if (!IsTupleType || (original._lazyTupleData == null && this._lazyTupleData == null) || TupleData!.EqualsIgnoringTupleUnderlyingType(original.TupleData))
             {
-                // Ensure all Rest extensions are tuples
-                tupleCompatibleType = EnsureRestExtensionsAreTuples(tupleCompatibleType);
-
-                var extensionTupleElementTypes = tupleCompatibleType.TypeArgumentsWithAnnotationsNoUseSiteDiagnostics[RestPosition - 1].Type.TupleElementTypesWithAnnotations;
-                var typesBuilder = ArrayBuilder<TypeWithAnnotations>.GetInstance(RestPosition - 1 + extensionTupleElementTypes.Length);
-                typesBuilder.AddRange(tupleCompatibleType.TypeArgumentsWithAnnotationsNoUseSiteDiagnostics, RestPosition - 1);
-                typesBuilder.AddRange(extensionTupleElementTypes);
-                elementTypes = typesBuilder.ToImmutableAndFree();
-            }
-            else
-            {
-                elementTypes = tupleCompatibleType.TypeArgumentsWithAnnotationsNoUseSiteDiagnostics;
+                return this;
             }
 
-            return new TupleTypeSymbol(locations, tupleCompatibleType, elementLocations, elementNames, elementTypes, errorPositions);
+            return WithElementNames(original.TupleElementNames, original.TupleElementLocations, original.TupleErrorPositions, original.Locations);
         }
 
+        internal NamedTypeSymbol? TupleUnderlyingType
+            => this._lazyTupleData != null ? this.TupleData!.TupleUnderlyingType : (this.IsTupleType ? this : null);
+
         /// <summary>
-        /// Adjust the <paramref name="tupleCompatibleType"/> in such a way so that 
+        /// Adjust the <paramref name="tupleCompatibleType"/> in such a way so that
         /// all types used by Rest fields are tuples. Throughout the entire nesting chain.
         /// </summary>
         private static NamedTypeSymbol EnsureRestExtensionsAreTuples(NamedTypeSymbol tupleCompatibleType)
```

Comment:
>EnsureRestExtensionsAreTuples [](start = 39, length = 29)

This entire method appears to be a no-op.

Reply:
Good catch. Thanks :)

---
In reply to: [357367525](https://github.com/dotnet/roslyn/pull/39370#discussion_r357367525) [](ancestors = 357367525)